### ISSUE TITLE
logistic_regression should be  able to reconstruct regression coefficients for simple data cases 

### ISSUE THREAD

**Issue by Sandy4321** [cvxgrp/cvxpy#775]
**Describe the bug**
in 
https://www.cvxpy.org/examples/machine_learning/logistic_regression.html

cvxpy logistic regression does not reconstruct regression coefficients when gradient descent does

**To Reproduce**
A [minimal code example](https://stackoverflow.com/help/mcve) that reproduces the bug or unexpected behavior.
just run 
https://github.com/lazyprogrammer/machine_learning_examples/blob/master/logistic_regression_class/l1_regularization.py
to see that regression coefficients  can be reconstructed 

**Expected behavior**
add code from link above to create data as there were created
```
import cvxpy as cp
import numpy as np

random_seed = 123
np.random.seed(random_seed)

flag_use_easy_test = 1
# **********************************
def sigmoid(z):
    return 1/(1 + np.exp(-z))
N = 50
D = 50 #7# origianl value in l1_regularization.py is 50

# uniformly distributed numbers between -5, +5
X_new = (np.random.random((N, D)) - 0.5)*10
# X = (np.random.randn(N, D) - 0.5)*10

# true weights - only the first 3 dimensions of X affect Y
true_w = np.array([1, 0.5, -0.5] + [0]*(D - 3))

# generate Y - add noise with variance 0.5
Y_new = np.round(sigmoid(X_new.dot(true_w) + np.random.randn(N)*0.5))
# **********************************
n = 3 #20
m = 10#1000
density = 0.2
beta_true = np.random.randn(n)
idxs = np.random.choice(range(n), int((1-density)*n), replace=False)
for idx in idxs:
    beta_true[idx] = 0

sigma = 3
X = np.random.normal(0, 5, size=(m,n))
X[:, 0] = 1.0
Y = X @ beta_true + np.random.normal(0, sigma, size=m)
Y[Y > 0] = 1
Y[Y <= 0] = 0

X_test = np.random.normal(0, 5, size=(m, n))
X_test[:, 0] = 1.0
if 0 : #origianl
    if 0 :
        Y_test = X_test @ beta_true + np.random.normal(0, sigma, size=(m,1))
    else:
        Y_test_wo_noise = X_test @ beta_true
        noise = np.random.normal(0, sigma, size=(m,1))
        Y_test = Y_test_wo_noise + noise 
        q=0
        
else: #changed to implement Y_test = X_test @ beta_true + np.random.normal(0, sigma, size=(m))
    if 1 :
        Y_test = X_test @ beta_true + np.random.normal(0, sigma, size=(m))
    else:
        Y_test_wo_noise = X_test @ beta_true
        noise = np.random.normal(0, sigma, size=(m))
        Y_test = Y_test_wo_noise + noise


Y_test[Y_test > 0] = 1
Y_test[Y_test <= 0] = 0

if flag_use_easy_test:
    n= D
    m = N
    Y_test = Y_new
    print('old Y shape ', Y.shape)
    Y = Y_new
    print('new Y shape ', Y.shape)
    X_test = X_new
    print('old X shape ', X.shape)
    X = X_new
    print('new X shape ', X.shape)
    


beta = cp.Variable(n)
lambd = cp.Parameter(nonneg=True)
log_likelihood = cp.sum(
    cp.multiply(Y, X @ beta) -
    cp.log_sum_exp(cp.vstack([np.zeros(m), X @ beta]), axis=0) -
    lambd * cp.norm(beta[1:], 1)
)
problem = cp.Problem(cp.Maximize(log_likelihood))

def error(scores, labels):
    scores[scores > 0] = 1
    scores[scores <= 0] = 0
    return np.sum(np.abs(scores - labels)) / float(np.size(labels))
trials = 10
train_error = np.zeros(trials)
test_error = np.zeros(trials)
lambda_vals = np.logspace(-2, 0, trials)
beta_vals = []
for i in range(trials):
    lambd.value = lambda_vals[i]
    problem.solve()
    train_error[i] = error(X @ beta.value, Y)
    test_error[i] = error(X_test @ beta.value, Y_test)
    beta_vals.append(beta.value)
    print('random_seed = ' , random_seed ,'lambd.value = ' ,  lambd.value, 'train_error[i] = ', train_error[i] )
    print(beta.value[0:10])
    print(' ')
    q=0

q=0
``` 

**Output**
If applicable, include program output. If reporting a program crash, please include the entire stack trace.

**Version**
 - OS: Windows
 - CVXPY Version: latest

**Additional context**

**Comment by jinglejiang** [comment]
beta.value
gives you regression coefficients

**Comment by SteveDiamond** [comment]
@jinglejiang the concern was more that in the actual example code, logistic regression does not do a good job recovering the true underlying beta. So you would need to fix the actual example so the beta values returned by cvxpy are close to the true beta. You can do this by removing the L1 norm penalty I suspect.

**Comment by Sandy4321** [comment]
good point 
but only please do not remove L1 norm
by the way 
https://github.com/scikit-learn-contrib/lightning
has l1 norm and dual and primal
you can use this to fix you code

**Comment by SteveDiamond** [comment]
@Sandy4321 I apologize for not taking your concerns seriously. I have fixed the example. There was a second bug in it. Let me know if you are satisfied.

**Comment by Sandy4321** [comment]
Great did you tested by test case from
https://github.com/vene/pyowl
?

**Comment by SteveDiamond** [comment]
The test case you sent is for solving a Lasso problem, which is totally different from L1-regularized logistic regression. I'll respond when you identify bugs, but I'm not going to do random work for you.

**Comment by Sandy4321** [comment]
At least can you kindly share link to read that lasso is not L1 regularisation

**Comment by Sandy4321** [comment]
As I can see from
https://www.google.com/url?sa=t&source=web&rct=j&url=https://towardsdatascience.com/l1-and-l2-regularization-methods-ce25e7fc831c&ved=2ahUKEwitv7OQwKrkAhXmYd8KHdu6BeYQFjACegQIEBAJ&usg=AOvVaw3Q_HZeIdpo00YLXUWbRGg6

regression model that uses L1 regularization technique is called Lasso Regression and model which uses L2 is called Ridge Regression.

**Comment by Sandy4321** [comment]
Or this

https://stats.stackexchange.com/questions/200416/is-regression-with-l1-regularization-the-same-as-lasso-and-with-l2-regularizati

**Comment by Sandy4321** [comment]
Or maybe you mean that lasso is for regression but L1 for logistics regression
But people use lasso term for logistics regression
https://www.google.com/amp/s/eight2late.wordpress.com/2017/07/11/a-gentle-introduction-to-logistic-regression-and-lasso-regularisation-using-r/amp/

**Comment by Sandy4321** [comment]
Or
http://www.sthda.com/english/articles/36-classification-methods-essentials/149-penalized-logistic-regression-essentials-in-r-ridge-lasso-and-elastic-net/
